### ISSUE TITLE
enable-remote-node-identity applies to KNP / k8s (CNI neutral) NetworkPolicy

### ISSUE THREAD

**Issue by agilgur5** [cilium/cilium#14989]
## Bug report

In Cilium 1.7, the [`enable-remote-node-identity` option](https://docs.cilium.io/en/v1.7/install/upgrade/#configmap-remote-node-identity) was created, and, when `true`, this requires that `remote-node` connectivity be explicitly allowed in NetworkPolicy.

Somewhat problematically, the CNI neutral k8s `NetworkPolicy` cannot specify `remote-node` as it is a `CiliumNetworkPolicy`-specific entity. Reading [the k8s `NetworkPolicy` docs](https://kubernetes.io/docs/concepts/services-networking/network-policies/#what-you-can-t-do-with-network-policies-at-least-not-yet), it sounds like this behavior is unsupported with plain, CNI neutral `NetworkPolicy` as node identities are currently unsupported.

Maybe I'm misreading the k8s docs, but as such, I was somewhat surprised that `enable-remote-node-identity` actually affected `NetworkPolicy` and not just `CiliumNetworkPolicy`, particularly given that `remote-node` and node identities in general can't be specified there.
I'd thought that `enable-remote-node-identity` would have no impact on KNP and that the KNP -> CNP translation code would handle adding `fromEntities: - remote-node` automatically, but it didn't.

### Specific Details

When I set `enable-remote-node-identity` to `true`, some of my pods, primarily those that used `hostNetwork: true`, started dropping connections as our `NetworkPolicy` only has `podSelectors` and so ended up denying `remote-node`.
Some specific examples include KIAM agents accessing the KIAM server via `hostNetwork` and `metrics-server` and KEDA receiving requests from the host's `kupe-apiserver` to serve metrics in the appropriate APIs. Or DNS as the example below.
I was only able to resolve this using CNP, despite starting out with only KNPs.

This is very related to https://github.com/kubernetes/kops/issues/10836, as the recently released kOps 1.19 seems to accidentally enable this option by default (and therefore accidentally created a breaking change during upgrade from kOps 1.18 to kOps 1.19), but I was still surprised it impacted KNP, not solely CNP

I'm not 100% sure that this is a "bug"/incorrect behavior, as it's not entirely clear to me what the behavior should be based on the k8s spec. This case seems like it might be undefined -- perhaps the appropriate resolution to this is better docs on the spec for k8s and Cilium. In any case, if this is incorrect behavior, it's a "spec interpretation issue" as opposed to an actual code bug per se.

**General Information**

- Cilium version (run `cilium version`)
  - This affects all versions since introduction of `enable-remote-node-identity`, so 1.7+
- Kernel version (run `uname -a`)
  - Amazon Linux 2, so kernel version 4.14
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)
  - k8s 1.19.7
  - kOps 1.19.0
- Link to relevant artifacts (policies, deployments scripts, ...)
  - see steps to reproduce below
- Generate and upload a system zip:
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```
  - N/A as this is a spec issue, not a bug in some code
  
**How to reproduce the issue**

Per https://github.com/kubernetes/kops/issues/10836, the simplest reproduction is:

1. Install Cilium with `enable-remote-node-identity` set to `true` (e.g. use kOps 1.19.0 with `kops create-cluster --networking cilium`)
2. In a cluster with `kube-dns` in the `kube-system` namespace, add a simple KNP:

```yaml
# Allow traffic from other namespaces to kube-dns (CoreDNS)
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: kube-dns
  namespace: kube-system
spec:
  podSelector:
    matchLabels:
      k8s-app: kube-dns
  ingress:
  - from:
    - namespaceSelector: {}
```

3. Send some request _from a host_ (e.g. `hostNetwork: true` pod or SSH'ed into a host) to `kube-dns`, e.g. `curl kube-dns.kube-system.cluster.service.local` or an `nslookup` to something. This request will fail and timeout
    - This is just a simple example using a service that already exists in a default cluster; any other service works too. I'm not totally sure exactly how host DNS works, this was just one service we had issues with.
    
4. This issue can then be resolved by replacing with an equivalent CNP that also includes `remote-node` (or, in this case, `cluster`, which includes `remote-node`):

```yaml
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: kube-dns-from-host
  namespace: kube-system
spec:
  endpointSelector:
    matchLabels:
      k8s-app: kube-dns
  ingress:
    - fromEntities:
      - cluster
```

A more complex reproduction would be to:

1. Deploy KIAM, `metrics-server`, or KEDA, or another service that uses `hostNetwork: true`, as mentioned above.
2. Then one can add a different, simple KNP:

```yaml
---
# Default network policy to deny all traffic from outside the namespace
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: default
  namespace: test-namespace-where-service-was-deployed
spec:
  podSelector:
    matchLabels:
  ingress:
  - from:
    - podSelector: {}
```

3. KIAM agents or `metrics-server` requests or KEDA requests etc will fail as they are made from the host, not from a pod.

4. This issue can then be resolved by adding a CNP that includes `remote-node`:

```yaml
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: from-hosts
  namespace: test-namespace-where-service-was-deployed
spec:
  endpointSelector:
    matchLabels:
      app: kiam-server
  ingress:
    - fromEntities:
      - host 
      - remote-node
```

As you can see from these examples, although I started with just KNPs, I required CNPs to resolve the denials caused by `enable-remote-node-identity`. Other than specifying the IP addresses or CIDR of the cluster's nodes, I'm not sure there's a simple way of achieving this in KNP as it does not currently have support for node identities. I'm also not sure that other CNIs behave the same way with respect to remote nodes and KNP.

### Can try to make a PR

If this behavior is indeed incorrect, I can try to make a PR to change the KNP -> CNP translation. I believe the code for this is located in [this block](https://github.com/cilium/cilium/blob/b875c1d61ba8e52c5a7e31d9f25165a4a9916d3e/pkg/k8s/network_policy.go#L141), though may need a bit more direction on which cases would be appropriate to add `fromEntities: - remote-node`, or if it should just be added to all to be backward-compatible with the previous behavior.

**Comment by joestringer** [comment]
Hi @agilgur5 , firstly thanks for the great writeup. This is indeed a known limitation with respect to the `enable-remote-node-identity` option. For a bit more context, there's another variation on this issue where I had posted some thoughts, those notes might also help you to reason about Cilium's behaviour here and maybe see about next steps: https://github.com/cilium/cilium/issues/12277 

I'm not sure I fully understand the proposal for KNP -> CNP translation. As far as I can tell, KNP is ambiguous on the point of how node traffic should be handled by the policy. The danger here is that if we are overly broad in our interpretation, we may end up opening up more traffic than the user really intends, particularly if you may have one view on how the policy should react and other users have a different view. The "nice" thing today about the `fromEntities: remote-node` approach in CNP is that the semantics are well-defined, and if the user wants pod->node traffic to be allowed, then it's very explicit in the policy so Cilium is definitely doing what they intend.  I think that what you're suggesting is to automatically inject an "allow to host/remote-node" rule based on certain KNP rule inputs, particularly `podSelector`. The thing is, if the pod selected by that pod selector has `hostNetwork: true`, then does that mean that any application directly running on that node should be accessible, even if not otherwise explicitly allowed by the policy? That's the behaviour it will have. Doing so from a `podSelector: {}` rule does not make this very obvious to users, and the nature of security issues is that they lurk in areas of implicit behaviour. Furthermore, I don't think we should encourage `podSelector: {}` rules either, as they are overly broad and not in line with the "least privilege" principle.

An alternative solution to the problem is to resolve #12277 and say that if KNP users want to allow traffic towards remote nodes, they must specify node IPs explicitly. I'm not the biggest fan of this approach either because I don't think that IP addressing really has a place in Kubernetes policies given that there are better higher-level constructs (and CNP provides these!), but that approach would at least provide a path for users who are not willing to use CNP, and in a way that would be a bit more explicit to the policy writer.

**Comment by agilgur5** [comment]
Thanks for the detailed response @joestringer . My apologies for not getting back to you in a more timely fashion; we ended up working around it by just completely porting over to CNP usage and then I got pulled into lots of things and left this at the wayside.

> As far as I can tell, KNP is ambiguous on the point of how node traffic should be handled by the policy.

Yea, KNP still [doesn't seem to support](https://kubernetes.io/docs/concepts/services-networking/network-policies/#what-you-can-t-do-with-network-policies-at-least-not-yet) node traffic outside of using CIDRs. In my opening comment, I also thought that this case was likely "undefined" in the KNP spec

> I'm not sure I fully understand the proposal for KNP -> CNP translation

Given that, my proposal was either to add more docs around this edge-case when using KNPs, or, given the ambiguity, to always allow `remote-node` traffic for KNPs as it can't be specified. Basically just for KNPs to be backward-compatible prior to the introduction of `enable-remote-node-identity`.

Now that this is like a year out though, changing that setting would likely be even more breaking, so that's probably not a viable option. Would have to check if any docs would still be useful a year out.

> The danger here is that if we are overly broad in our interpretation, we may end up opening up more traffic than the user really intends

Totally agreed, one of the reasons I'm a Cilium stan is because of lots of great security, least-privilege, and explicitness decisions and implementations like this!
Unfortunately KNPs and other implementations tend to lag behind, which can create some compatibility issues due to non-existent features such as this.

> I think that what you're suggesting is to automatically inject an "allow to host/remote-node" rule based on certain KNP rule inputs

I actually didn't mean something as complex as that, and meant the very simple backward-compatibility default I mentioned above 😅  (but I'll take that as a compliment haha). 
It would be super cool if there were a way to automatically figure that out, though. I tried wrapping my head around it for a bit after you first replied and couldn't come up with a good strategy that wouldn't be confusing, as you mentioned

> I'm not the biggest fan of this approach either because I don't think that IP addressing really has a place in Kubernetes policies given that there are better higher-level constructs (and CNP provides these!), but that approach would at least provide a path for users who are not willing to use CNP, and in a way that would be a bit more explicit to the policy writer.

Also agreed. At least in my case, we probably couldn't use that solution as we don't generally rely on sticky IPs. But may be useful to others as a stop-gap 🤷
